File: nlp_service/app/schemas/blueprint.py
"""
Pydantic schemas for POST /api/nlp/generate-blueprint.

`UnitWeightIn.weightage` plays the role of spec Section 13's
BLUEPRINTS.topicDistribution field, applied at syllabus-unit granularity
(matching Phase 2's `units: [{unitNumber, title, topics}]` output)
rather than a free-text topic list.
"""

from __future__ import annotations

from typing import Dict, List, Optional

from pydantic import BaseModel, Field


class UnitWeightIn(BaseModel):
    unitNumber: int
    title: Optional[str] = None
    weightage: float


class GenerateBlueprintRequest(BaseModel):
    totalMarks: float
    totalQuestions: int
    units: List[UnitWeightIn]
    difficultyDistribution: Dict[str, float]
    bloomDistribution: Dict[str, float]


class UnitBandOut(BaseModel):
    unitNumber: int
    title: str
    percentage: float
    questionCount: int
    marks: float


class BandOut(BaseModel):
    label: str
    percentage: float
    questionCount: int
    marks: float


class GenerateBlueprintResponseData(BaseModel):
    totalMarks: float
    totalQuestions: int
    units: List[UnitBandOut]
    difficulty: List[BandOut]
    bloom: List[BandOut]
    warnings: List[str] = Field(default_factory=list)

File: server/models/Blueprint.js
import mongoose from 'mongoose';

const bandSchema = new mongoose.Schema(
  { label: String, percentage: Number, questionCount: Number, marks: Number },
  { _id: false }
);

const unitBandSchema = new mongoose.Schema(
  { unitNumber: Number, title: String, percentage: Number, questionCount: Number, marks: Number },
  { _id: false }
);

const blueprintSchema = new mongoose.Schema(
  {
    teacherId: { type: mongoose.Schema.Types.ObjectId, ref: 'User', default: null },
    syllabusId: { type: mongoose.Schema.Types.ObjectId, ref: 'Syllabus', default: null },
    totalMarks: { type: Number, required: true },
    totalQuestions: { type: Number, required: true },
    units: { type: [unitBandSchema], default: [] },
    difficulty: { type: [bandSchema], default: [] },
    bloom: { type: [bandSchema], default: [] },
    warnings: { type: [String], default: [] },
  },
  { timestamps: true }
);

export default mongoose.models.Blueprint || mongoose.model('Blueprint', blueprintSchema);
